### PR TITLE
Implement HTTP header/cookie capture APIs

### PR DESCRIPTION
**Pull request by Miciah**
#### Implement HTTP header capture API

This change is related to [NE-320](https://issues.redhat.com/browse/NE-320).

* `pkg/operator/controller/ingress/deployment.go` (`RouterCaptureHTTPRequestHeaders`, `RouterCaptureHTTPResponseHeaders`): New constants with the names of the related environment variables.
(`desiredRouterDeployment`): Set the `ROUTER_CAPTURE_HTTP_REQUEST_HEADERS` and `ROUTER_CAPTURE_HTTP_RESPONSE_HEADERS` environment variables as appropriate.
(`accessLoggingForIngressController`): Set `HTTPCaptureHeaders`.
(`serializeCaptureHeaders`): New function.  Serialize HTTP header capture configuration into values suitable for the environment variables.
* `pkg/operator/controller/ingress/deployment_test.go` (`TestDesiredRouterDeployment`): Verify that `spec.logging.access.httpCaptureHeaders` has the expected effect.
* `test/e2e/operator_test.go` (`TestHTTPHeaderCapture`): New test.

#### Implement HTTP cookie capture API

* `pkg/operator/controller/ingress/deployment.go` (`RouterCaptureHTTPCookies`): New constant with the name of the related environment variable.
(`desiredRouterDeployment`): Set the `ROUTER_CAPTURE_HTTP_COOKIE` environment variable as appropriate.
(`accessLoggingForIngressController`): Set `HTTPCaptureCookies`.
* `pkg/operator/controller/ingress/deployment_test.go` (`TestDesiredRouterDeployment`): Verify that `spec.logging.access.httpCaptureCookies` has the expected effect.
* `test/e2e/operator_test.go` (`TestHTTPCookieCapture`): New test.

---

Implements the following enhancement: https://github.com/openshift/enhancements/pull/370

Depends on the following router changes: https://github.com/openshift/router/pull/139, https://github.com/openshift/router/pull/147

Depends on the following API changes: https://github.com/openshift/api/pull/686. https://github.com/openshift/api/pull/687

### CODE REVIEW

**Comment by Miciah** [comment]
e2e-aws-operator failed with the following:

    Cluster operator console Degraded is True with RouteHealth_StatusError: RouteHealthDegraded: route not yet available, https://console-openshift-console.apps.ci-op-g8lrd4lv-43abb.origin-ci-int-aws.dev.rhcloud.com/health returns '503 Service Unavailable'

and

    Cluster operator monitoring Degraded is True with UpdatingkubeStateMetricsFailed: Failed to rollout the stack. Error: running task Updating kube-state-metrics failed: reconciling kube-state-metrics Deployment failed: updating Deployment object failed: waiting for DeploymentRollout of kube-state-metrics: deployment kube-state-metrics is not ready. status: (replicas: 1, updated: 1, ready: 1, unavailable: 0)

The console pod is logging the following error:

    auth: error contacting auth provider (retrying in 10s): discovery through endpoint https://kubernetes.default.svc/.well-known/oauth-authorization-server failed: 404 Not Found

So I'm not sure whether that is caused by the changes in this PR.  Let's see whether it happens again.

/test e2e-aws-operator

File: pkg/operator/controller/ingress/deployment.go [hunk 1]
```diff
@@ -463,6 +466,16 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
 		}
 	}
 
+	if ci.Spec.Logging != nil && ci.Spec.Logging.Access != nil && ci.Spec.Logging.Access.HTTPCaptureHeaders != nil {
+		env = append(env, corev1.EnvVar{
+			Name:  RouterCaptureHTTPRequestHeaders,
+			Value: serializeCaptureHeaders(ci.Spec.Logging.Access.HTTPCaptureHeaders.Request),
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment.go:472]: Is there a guarantee that if `ci.Spec.Logging.Access.HTTPCaptureHeaders` is non-nil, both `ci.Spec.Logging.Access.HTTPCaptureHeaders.Request` and `ci.Spec.Logging.Access.HTTPCaptureHeaders.Response` are also non-nil? What if `ci.Spec.Logging.Access.HTTPCaptureHeaders.Request` were to be nil, but `ci.Spec.Logging.Access.HTTPCaptureHeaders.Response` was non-nil?

Comment (Miciah) [reply pkg/operator/controller/ingress/deployment.go:472]: > Is there a guarantee that if `ci.Spec.Logging.Access.HTTPCaptureHeaders` is non-nil, both `ci.Spec.Logging.Access.HTTPCaptureHeaders.Request` and `ci.Spec.Logging.Access.HTTPCaptureHeaders.Response` are also non-nil?

No.  It is possible to specify `httpCaptureHeaders` with nil `request` and `response` values.  In this case, the operator will specify empty values for `ROUTER_CAPTURE_HTTP_REQUEST_HEADERS` and `ROUTER_CAPTURE_HTTP_RESPONSE_HEADERS`, which is harmless.

> What if `ci.Spec.Logging.Access.HTTPCaptureHeaders.Request` were to be nil, but `ci.Spec.Logging.Access.HTTPCaptureHeaders.Response` was non-nil?

Then `ROUTER_CAPTURE_HTTP_REQUEST_HEADERS` would be empty and `ROUTER_CAPTURE_HTTP_RESPONSE_HEADERS` would be nonempty.  This should be fine; the `range` expressions that https://github.com/openshift/router/pull/139/files adds will iterate over the 0 comma-delimited values and add 0 `capture` stanzas.

Comment (sgreene570) [reply pkg/operator/controller/ingress/deployment.go:472]: > the operator will specify empty values for `ROUTER_CAPTURE_HTTP_REQUEST_HEADERS` and `ROUTER_CAPTURE_HTTP_RESPONSE_HEADERS`, which is harmless.

Ah, ok, wasn't sure if empty values were intended or not. Thanks!

**Comment by sgreene570** [comment]
Just reviewed this again. e2e & unit test coverage looks great :smiley: 
/lgtm

**Comment by Miciah** [comment]
[The last failure](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/409/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws/1288522971937247232) looks like [BZ1860837 – Failing test [It] should provide named network metrics [Suite:openshift/conformance/parallel]](https://bugzilla.redhat.com/show_bug.cgi?id=1860837).

**Comment by Miciah** [comment]
Also seeing [BZ1861201 – [sig-cli] oc adm must-gather runs successfully for audit logs](https://bugzilla.redhat.com/show_bug.cgi?id=1861201) in [1288362043102990336](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/409/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws/1288362043102990336).

**Pull request merged**